### PR TITLE
fix: fix altair api break, cleanup DependencyManager

### LINKED ISSUE #2005: Possible incompatibility when using marimo_csv transformer with altair 5.4.0
### Describe the bug

When generating a plot using the marimo_csv transformer, I'm getting a DataFrame constructor not properly called error.

This seems to be linked to an upgrade from altair `5.3.0` to `5.4.0` on my end, but only happens with the marimo csv transformer as far as I can tell.

```
Traceback (most recent call last):
  File "/Users/andrewhill/Desktop/absel/env/lib/python3.10/site-packages/marimo/_output/formatting.py", line 172, in try_format
    mimetype, data = formatter(obj)
  File "/Users/andrewhill/Desktop/absel/env/lib/python3.10/site-packages/marimo/_output/formatters/altair_formatters.py", line 44, in _show_chart
    srcdoc=html.escape(chart.to_html()),
  File "/Users/andrewhill/Desktop/absel/env/lib/python3.10/site-packages/altair/vegalite/v5/api.py", line 1946, in to_html
    self.to_dict(),
  File "/Users/andrewhill/Desktop/absel/env/lib/python3.10/site-packages/altair/vegalite/v5/api.py", line 3772, in to_dict
    return super().to_dict(**kwds)
  File "/Users/andrewhill/Desktop/absel/env/lib/python3.10/site-packages/altair/vegalite/v5/api.py", line 1798, in to_dict
    copy.data = _prepare_data(data, context)
  File "/Users/andrewhill/Desktop/absel/env/lib/python3.10/site-packages/altair/vegalite/v5/api.py", line 275, in _prepare_data
    data = func(data)
  File "/Users/andrewhill/Desktop/absel/env/lib/python3.10/site-packages/marimo/_plugins/ui/_impl/charts/altair_transformer.py", line 56, in _to_marimo_csv
    data_csv = _data_to_csv_string(data)
  File "/Users/andrewhill/Desktop/absel/env/lib/python3.10/site-packages/marimo/_plugins/ui/_impl/charts/altair_transformer.py", line 99, in _data_to_csv_string
    return get_table_manager(data).to_csv().decode("utf-8")
  File "/Users/andrewhill/Desktop/absel/env/lib/python3.10/site-packages/marimo/_plugins/ui/_impl/tables/default_table.py", line 89, in to_csv
    return self._as_table_manager().to_csv(format_mapping)
  File "/Users/andrewhill/Desktop/absel/env/lib/python3.10/site-packages/marimo/_plugins/ui/_impl/tables/default_table.py", line 161, in _as_table_manager
    return PandasTableManagerFactory.create()(pd.DataFrame(self.data))
  File "/Users/andrewhill/Desktop/absel/env/lib/python3.10/site-packages/pandas/core/frame.py", line 886, in __init__
    raise ValueError("DataFrame constructor not properly called!")
ValueError: DataFrame constructor not properly called!
```

### Environment

```
{
  "marimo": "0.7.19",
  "OS": "Darwin",
  "OS Version": "23.5.0",
  "Processor": "arm",
  "Python Version": "3.10.4",
  "Binaries": {
    "Browser": "--",
    "Node": "v21.7.1"
  },
  "Requirements": {
    "click": "8.1.3",
    "importlib-resources": "missing",
    "jedi": "0.19.1",
    "markdown": "3.6",
    "pymdown-extensions": "10.8.1",
    "pygments": "2.17.2",
    "tomlkit": "0.12.4",
    "uvicorn": "0.29.0",
    "starlette": "0.37.2",
    "websockets": "12.0",
    "typing-extensions": "4.11.0",
    "ruff": "0.3.7"
  }
}
```

### Code to reproduce

```
@app.cell
def __():
    import marimo as mo
    import polars as pl
    import altair as alt

   # if remove this line and restart kernel, plot appears as expected
    alt.data_transformers.enable("marimo_csv")

    test_data = pl.DataFrame({
        "name": ["Alice", "Bob", "Charlie", "David"],
        "salary": [50000, 60000, 75000, 55000],
    })


    (
        alt.Chart(test_data.to_pandas())
        .encode(
            y="salary",
            x="name",
        )
        .mark_point(color="red")
    )

    return alt, mo, pl, test_data


if __name__ == "__main__":
    app.run()
```

### PR DESCRIPTION
**Pull request by mscolnick**
Fixes #2005 

Altair made an (arguably) breaking change. Their data transformers now pass narwhal dataframes instead of pandas, which breaks some downstream features in marimo. This PR now supports [narwhals](https://github.com/narwhals-dev/narwhals)

This is nice actually nice since now we can pass polars, pyarrow, etc to altair charts and get back the polars, pyarrow, etc charts.  

This also DRYs up `DependencyManager`

### CODE REVIEW

File: marimo/_ast/visitor.py [hunk 1]
```diff
@@ -365,7 +365,7 @@ def visit_Call(self, node: ast.Call) -> None:
             elif isinstance(first_arg, ast.JoinedStr):
                 sql = normalize_sql_f_string(first_arg)
 
-            if isinstance(sql, str) and DependencyManager.has_duckdb() and sql:
+            if isinstance(sql, str) and DependencyManager.duckdb.has() and sql:
```
Comment (akshayka) [inline marimo/_ast/visitor.py:368]: Should this be `has_at_version`? `extract_statements` isn't available in some earlier versions of `duckdb`.

File: marimo/_sql/sql.py [hunk 2]
```diff
@@ -33,7 +33,7 @@ def sql(
     Returns:
         The result of the query.
     """
-    DependencyManager.require_duckdb("to execute sql")
+    DependencyManager.duckdb.require("to execute sql")
```
Comment (akshayka) [inline marimo/_sql/sql.py:36]: Should this be `require_at_version`?

File: marimo/_dependencies/dependencies.py [hunk 3]
```diff
@@ -1,203 +1,143 @@
...
+        return _version_check(
+            pkg=self.pkg,
+            v=self.get_version(),
+            min_v=min_version,
+            max_v=max_version,
+        )
 
-    @staticmethod
-    def require_numpy(why: str) -> None:
+    def require(self, why: str) -> None:
```
Comment (akshayka) [inline marimo/_dependencies/dependencies.py:43]: Should we also have a `require_at_version`?

File: marimo/_ast/visitor.py (hunk 1 above)
Comment (mscolnick) [reply marimo/_ast/visitor.py:368]: yea i can add it here too

File: marimo/_dependencies/dependencies.py (hunk 3 above)
Comment (mscolnick) [reply marimo/_dependencies/dependencies.py:43]: i can add it, but I dont think it is used yet

File: marimo/_dependencies/dependencies.py [hunk 4]
```diff
@@ -1,203 +1,160 @@
...
-        if not DependencyManager.has_pyarrow():
-            raise ModuleNotFoundError(
-                f"pyarrow is required {why}. "
-                + "You can install it with 'pip install pyarrow'"
-            ) from None
+    pandas = Dependency("pandas")
+    polars = Dependency("polars")
+    numpy = Dependency("numpy")
+    altair = Dependency("altair", min_version="5.3.0", max_version="6.0.0")
+    duckdb = Dependency("duckdb")
```
Comment (akshayka) [inline marimo/_dependencies/dependencies.py:143]: `min_version=1.0.0`? Or do we support `duckdb < 1.0` in some places?

**Comment by MarcoGorelli** [comment]
Hey, sorry to chime in, just wanted to
1. say hi, I'm a fan of Marimo 😍 
2. apologise for having introduced the breaking change (I was worried that at least _something_ would happen... 🥶 )
3. invite you to please feel free to reach out if you need or would like any assistance with Narwhals
4. understand whether this needs patching in Altair. From your comment "This is nice actually nice since now we can pass polars, pyarrow, etc to altair charts and get back the polars, pyarrow, etc charts." I think you're OK with the change, but it just should have been called out more explicitly as breaking?

Thanks 🙏

**Review by akshayka — APPROVED** [review]

**Comment by mscolnick** [comment]
Hi @MarcoGorelli - thanks for chiming in (and being a fan!)

2. No worries. I'm not sure how to make this backward compatible - could just pass the native DF instead of narwhals' DF. Speaking for marimo, its fine for us to fix forward. But I am not sure how much this affects other downstream projects.
4. I will definitely reach out, thank you for offering. 
5. Yea maybe a callout (although I found out first from a user instead of the changelog). This is a better change longer term for marimo. I guess in future, we can collaborate on RC releases to make sure they work with marimo.

**Comment by MarcoGorelli** [comment]
just fyi, the next Altair version should address this https://github.com/vega/altair/pull/3550

**Comment by mscolnick** [comment]
thank you @MarcoGorelli!

**Pull request merged**